### PR TITLE
MS19054: [CreateApp/EntityList] auto scroll

### PR DESCRIPTION
**Pull request by thoys**
https://highfidelity.manuscript.com/f/cases/19054/Selecting-an-item-should-scroll-the-entity-list-to-show-the-item

## QA Test

1. Open the create app
2. Create more entities than fit in the viewable area of the list, if there are not too many entities yet
3. Select entities in the 3d window. (try at least 10 different ones) .
       - The list should always move with the least amount of effort, if the entity was above the viewable area, the entity would be on top, and vice versa.
       - When the entity is already fully in the viewable area of the list then the list should not scroll.
4. Select multiple entities in the 3d window using shift. (try a couple of times)
       - The top of the list entity should always be in view.
       - The bottom selected entity in the list will only be visible when it fits.
       - In case that there is a bottom selected entity in the list, the first entity should be on top.

5. Smoke test the entity list.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13679/HighFidelity-Beta-PR14149-7713fb6.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13679/HighFidelity-Beta-PR14149-7713fb6.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13679/HighFidelity-Beta-Interface-PR14149-7713fb6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13679/HighFidelity-Beta-PR14149-7713fb6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13679/HighFidelity-Beta-Interface-PR14149-7713fb6.exe

File: scripts/system/html/js/entityList.js [hunk 1]
```diff
@@ -193,7 +193,7 @@ function loaded() {
                 if (selectedIndex >= 0) {
                     selection = [];
                     selection = selection.concat(selectedEntities);
-                    selection.splice(selectedIndex, 1)
+                    selection.splice(selectedIndex, 1);
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/js/entityList.js]: I agree, the semicolon makes it look like a ); in the end

File: scripts/system/html/js/listView.js [hunk 2]
```diff
@@ -152,6 +152,24 @@ ListView.prototype = {
...
+        let boundingTop = rowIndex * this.rowHeight;
+        let boundingBottom = (lastRowIndex * this.rowHeight) + this.rowHeight;
+        if ((boundingBottom - boundingTop) > this.elTableScroll.clientHeight) {
+            boundingBottom = boundingTop + this.elTableScroll.clientHeight;
+        }
+
+        let currentVisibleAreaTop = this.elTableScroll.scrollTop;
+        let currentVisibleAreaBottom = currentVisibleAreaTop + this.elTableScroll.clientHeight;
+
+        if (boundingTop < currentVisibleAreaTop) {
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/js/listView.js]: I believe it calls the scroll event when we change this value, which fixes that.

File: scripts/system/libraries/entityList.js [hunk 3]
```diff
@@ -224,7 +225,7 @@ EntityListTool = function(shouldUseEditTabletApp) {
             for (var i = 0; i < ids.length; i++) {
                 entityIDs.push(ids[i]);
             }
-            selectionManager.setSelections(entityIDs);
+            selectionManager.setSelections(entityIDs, "entityList");
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/libraries/entityList.js]: sure, but I'm transmitting it back to HTML which probably converts it to the string `[Object]`. I could probably skip sending it to html when intercepting it. It doesn't add extra value in sending the event to html when we're going to ignore it. Good one.

File: scripts/system/html/js/listView.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (dback2) [reply scripts/system/html/js/listView.js]: Ah ok that makes sense

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13726/HighFidelity-Beta-PR14149-739bdc9.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13726/HighFidelity-Beta-PR14149-739bdc9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13726/HighFidelity-Beta-Interface-PR14149-739bdc9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13726/HighFidelity-Beta-PR14149-739bdc9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13726/HighFidelity-Beta-Interface-PR14149-739bdc9.exe

**Comment by JeffClinton** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13993/HighFidelity-Beta-PR14149-739bdc9.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13993/HighFidelity-Beta-PR14149-739bdc9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13993/HighFidelity-Beta-Interface-PR14149-739bdc9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13993/HighFidelity-Beta-PR14149-739bdc9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13993/HighFidelity-Beta-Interface-PR14149-739bdc9.exe

**Comment by QliCknox** [comment]
Currently Testing

**Comment by QliCknox** [comment]
Pass @conklin94122 

Autoscrolling works as intended. It is smooth and always finds the quickest route. The smoke test around the entity list also passes.

**Comment by thoys** [comment]
Build this please.

**Comment by thoys** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14173/HighFidelity-Beta-PR14149-faaf12b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14173/HighFidelity-Beta-Interface-PR14149-faaf12b.dmg

**Comment by thoys** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14179/HighFidelity-Beta-PR14149-faaf12b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14179/HighFidelity-Beta-PR14149-faaf12b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14179/HighFidelity-Beta-Interface-PR14149-faaf12b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14179/HighFidelity-Beta-PR14149-faaf12b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14179/HighFidelity-Beta-Interface-PR14149-faaf12b.exe

**Comment by conklin94122** [comment]
@thoys QA Ready? Please label.

**Comment by thoys** [comment]
oh yeah sorry

**Comment by QliCknox** [comment]
Fail @conklin94122 

Retested and found that when you double click to focus - the entity is no longer selected in the entity list. This does occur in the last build I tested as well, 739bdc9

Repro:
1. Launch this PR
2. Load into any domain with entities
3. Double click on any  entity to 'Focus' 

Video and Log in .zip
[Archive 2.zip](https://github.com/highfidelity/hifi/files/2497105/Archive.2.zip)

This does not occur on the latest Master build, nor Stable (73.2)

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14261/HighFidelity-Beta-PR14149-f6b5f5e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14261/HighFidelity-Beta-PR14149-f6b5f5e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14261/HighFidelity-Beta-Interface-PR14149-f6b5f5e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14261/HighFidelity-Beta-PR14149-f6b5f5e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14261/HighFidelity-Beta-Interface-PR14149-f6b5f5e.exe

**Comment by qligmaloney** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14317/HighFidelity-Beta-PR14149-f6b5f5e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14317/HighFidelity-Beta-Interface-PR14149-f6b5f5e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14317/HighFidelity-Beta-PR14149-f6b5f5e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14317/HighFidelity-Beta-Interface-PR14149-f6b5f5e.exe

**Comment by qligmaloney** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14351/HighFidelity-Beta-PR14149-f6b5f5e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14351/HighFidelity-Beta-Interface-PR14149-f6b5f5e.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14353/HighFidelity-Beta-PR14149-5f35a32.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14353/HighFidelity-Beta-PR14149-5f35a32.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14353/HighFidelity-Beta-Interface-PR14149-5f35a32.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14353/HighFidelity-Beta-PR14149-5f35a32.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14353/HighFidelity-Beta-Interface-PR14149-5f35a32.exe

**Comment by QliCknox** [comment]
Currently Testing

**Pull request merged**